src/Bicep.Core/Emit/ResourceDependencyVisitor.cs

```diff
@@ -11,89 +11,42 @@ namespace Bicep.Core.Emit
 {
     public class ResourceDependencyVisitor : SyntaxVisitor
     {
+        // shared data structure across all ResourceDependencyVisitors; our only "state" during traversals
+        // Instead of traversing through variables repeatedly, we pre-calculate variable-resource dependencies before we build the graph
+        private static Dictionary<VariableSymbol, ImmutableHashSet<DeclaredSymbol>> variableDependencies = new Dictionary<VariableSymbol, ImmutableHashSet<DeclaredSymbol>>();
         private readonly SemanticModel model;
-        private IDictionary<DeclaredSymbol, HashSet<DeclaredSymbol>> resourceDependencies;
-        private DeclaredSymbol? currentDeclaration;
+        // resource dependencies specific to each ResourceDependencyVisitor
+        private HashSet<DeclaredSymbol> resourceDependencies;
 
-        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetResourceDependencies(SemanticModel model)
+        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetAllResourceDependencies(SemanticModel model) 
         {
-            var visitor = new ResourceDependencyVisitor(model);
-            visitor.Visit(model.Root.Syntax);
+            var allResourceDependencies = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
+            var declaredSymbols = model.Root.AllDeclarations.ToList(); // gets all declaredSymbols in the document (entry point)
 
-            var output = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
-            foreach (var kvp in visitor.resourceDependencies)
+            // we need to precompute variable dependencies to reuse them when traversing resources or modules
+            foreach(var variableSymbol in declaredSymbols.OfType<VariableSymbol>()) 
```

Comment:
You don't have to do this - `model.Root.VariableDeclarations` has all the computed the variable declarations ready for you to use.